crates/wasi-common/src/fdpool.rs

```diff
@@ -0,0 +1,132 @@
+use std::collections::{HashSet, VecDeque};
+use std::fmt;
+use std::hash::Hash;
+
+pub(crate) trait Fd: Copy + Clone + Hash + PartialEq + Eq + Default {
+    fn next(&self) -> Option<Self>;
+}
+
+pub(crate) struct FdPool<T: Fd> {
+    available: VecDeque<T>,
+    claimed: HashSet<T>,
+}
+
+impl<T: Fd + fmt::Debug> fmt::Debug for FdPool<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("FdPool")
+            .field("available", &self.available)
+            .field("claimed", &self.claimed)
+            .finish()
+    }
+}
+
+impl<T: Fd> FdPool<T> {
+    const BATCH_SIZE: usize = 64;
```

Comment:
I think we can probably avoid a literal list-of-indices here perhaps? We could have a `Vec` of available indices, as well as a "next index to hand out" stored as `Option<Fd>`. Allocation would pop from the `Vec` or otherwise `.take()` the next index. If the next index was taken we store the result of `.next()` back in there. 

Later during `deallocate` we'd just push the entry onto the `Vec`.

I think that may simplify a bit here and make it a bit more lightweight in terms of overhead?

Reply:
So basically, you suggest to skip the preallocation step, did I get it right? If so, this is what I originally had, just thought we could preallocate a couple upfront as means of optimisation but I guess that might have been somewhat premature ;-)

Reply:
Right yeah I think the preallocation can largely be skipped. We can always continue to optimize it more though if necessary!

Reply:
Agreed! Let me roll it back then!